### PR TITLE
feat: Add model_load_time metric

### PR DESCRIPTION
**Pull request by indrajit96**
#### What does the PR do?
Add new metric nv_load_time per model to metrics
Added load time gauge metric per model.
New metric added example

```
HELP nv_model_load_time Load Time per-model
TYPE nv_model_load_time gauge
nv_model_load_time{model="libtorch_float32_float32_float32",version="1"} 0.311423712
```

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [x] Related issues are referenced.
- [x] Added [test plan](#test-plan) and verified test passes.
- [ ] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [x] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [x] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Related PRs:
https://github.com/triton-inference-server/server/pull/7697

#### Where should the reviewer start?
ReportModelLoadTime func use-age.

#### Test plan:
Added in server PR

- CI Pipeline ID: https://gitlab-master.nvidia.com/dl/dgx/tritonserver/-/pipelines/19322776

### CODE REVIEW

File: src/model_repository_manager/model_lifecycle.cc [hunk 1]
```diff
@@ -526,6 +528,7 @@ ModelLifeCycle::AsyncLoad(
     load_pool_->Enqueue([this, model_id, version, model_info, OnComplete,
                          load_tracker, is_config_provided]() {
       for (size_t retry = 0; retry <= options_.load_retry; ++retry) {
+        // TODO add here
```
Comment (rmccorm4) [inline src/model_repository_manager/model_lifecycle.cc:531]: add what here?

File: src/model_repository_manager/model_lifecycle.cc [hunk 2]
```diff
@@ -847,4 +860,19 @@ ModelLifeCycle::OnLoadFinal(
...
 }
 
+void
+ModelLifeCycle::ReportModelLoadTime(
+    std::shared_ptr<MetricModelReporter> reporter,
+    const std::chrono::duration<double>& time_to_load)
+{
+#ifdef TRITON_ENABLE_METRICS
+  // Pending request count should always be 0 or 1 per-request. A request should
+  // not decrement the count unless it has already been incremented.
```
Comment (rmccorm4) [inline src/model_repository_manager/model_lifecycle.cc:870]: Is this comment related or copied?

File: src/metrics.cc [hunk 3]
```diff
@@ -109,6 +109,11 @@ Metrics::Metrics()
                     "execution per-model.")
               .Register(*registry_)),
 
+      model_load_time_family_(prometheus::BuildGauge()
+                                  .Name("nv_model_load_time")
+                                  .Help("Load Time per-model in nanoseconds")
```
Comment (rmccorm4) [inline src/metrics.cc:114]: nanoseconds or seconds?

File: src/model_repository_manager/model_lifecycle.cc (hunk 2 above)
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc]: Copied the usage of metrics reporter. Comment came along by mistake. Fixed.

File: src/model_repository_manager/model_lifecycle.cc (hunk 1 above)
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc]: Initially marked as a potential place to start the counter. Forgot to clean-up the comment.
Fixed

File: src/metrics.cc (hunk 3 above)
Comment (indrajit96) [reply src/metrics.cc]: Seconds.
Fixed it.

File: src/model_repository_manager/model_lifecycle.cc [hunk 4]
```diff
@@ -799,10 +802,20 @@ ModelLifeCycle::OnLoadFinal(
...
-
       loaded.second->state_ = ModelReadyState::READY;
       loaded.second->state_reason_.clear();
-
+      auto reporter = loaded.second->model_->MetricReporter();
+      const uint64_t now_ns =
+          std::chrono::duration_cast<std::chrono::nanoseconds>(
+              std::chrono::steady_clock::now().time_since_epoch())
+              .count();
+      uint64_t time_to_load_ns = now_ns - loaded.second->load_start_ns_;
```
Comment (rmccorm4) [inline src/model_repository_manager/model_lifecycle.cc:811]: should these be included in the `ifdef TRITON_ENABLE_METRICS` block?

File: src/metric_model_reporter.h [hunk 5]
```diff
@@ -88,6 +88,8 @@ class MetricModelReporter {
   // Lookup counter metric by name, and increment it by value if it exists.
   void IncrementCounter(const std::string& name, double value);
   // Increase gauge by value.
+  void SetGauge(const std::string& name, double value);
```
Comment (rmccorm4) [inline src/metric_model_reporter.h:91]: Fix comment from "increase gauge by value" to "overwrite gauge to value"

File: src/model_repository_manager/model_lifecycle.cc (hunk 4 above)
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:812]: Updated fdef TRITON_ENABLE_METRICS block.
Resolved

File: src/metric_model_reporter.h (hunk 5 above)
Comment (indrajit96) [reply src/metric_model_reporter.h:91]: Fixed.

File: src/metrics.cc [hunk 6]
```diff
@@ -109,6 +109,11 @@ Metrics::Metrics()
                     "execution per-model.")
               .Register(*registry_)),
 
+      model_load_time_family_(prometheus::BuildGauge()
+                                  .Name("nv_model_load_time")
```
Comment (rmccorm4) [inline src/metrics.cc:113]: See my comments on doc - curious if we need to match an exact metric name or not. For example, if the prefix `nv_` is OK or not.

**Comment by yinggeh** [comment]
Can you also add me to this PR?

File: src/model_repository_manager/model_lifecycle.cc [hunk 7]
```diff
@@ -799,10 +801,20 @@ ModelLifeCycle::OnLoadFinal(
...
+#ifdef TRITON_ENABLE_METRICS
+      auto reporter = loaded.second->model_->MetricReporter();
+      const uint64_t now_ns =
+          std::chrono::duration_cast<std::chrono::nanoseconds>(
+              std::chrono::steady_clock::now().time_since_epoch())
+              .count();
+      uint64_t time_to_load_ns = now_ns - loaded.second->load_start_ns_;
+      std::chrono::duration<double> time_to_load =
+          std::chrono::duration_cast<std::chrono::duration<double>>(
+              std::chrono::nanoseconds(time_to_load_ns));
```
Comment (GuanLuo) [inline src/model_repository_manager/model_lifecycle.cc:815]: Why don't you time the load time [here](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager/model_lifecycle.cc#L530) and directly storing the load time in model info. You can still put the metric update logic here.

File: src/metrics.cc (hunk 6 above)
Comment (yinggeh) [inline src/metrics.cc:113]: Let it be consistent with others, e.g. `nv_model_load_duration_s`.

File: src/metrics.cc [hunk 8]
```diff
@@ -109,6 +109,11 @@ Metrics::Metrics()
...
+                                  .Help("Load Time per-model in seconds")
```
Comment (yinggeh) [inline src/metrics.cc:114]: How about "Model loading duration in seconds"?
I don't think we need to explicitly say per-model as the metric label already tells.

File: src/model_repository_manager/model_lifecycle.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc]: Already removed in a previous commit.
Marking as resolved

File: src/model_repository_manager/model_lifecycle.cc (hunk 7 above)
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:815]: Reason I added in the metric update in 
`ModelLifeCycle::OnLoadFinal()`  was because that is where we set the status to READY for the model
`loaded.second->state_ = ModelReadyState::READY;`
If I set the model load time [here](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager/model_lifecycle.cc#L530) there is chance that `ModelLifeCycle::OnLoadFinal()` exits without loading the model and we might have a incorrect metric.

File: src/model_repository_manager/model_lifecycle.cc [hunk 9]
```diff
@@ -470,6 +471,7 @@ ModelLifeCycle::AsyncLoad(
     ModelInfo* model_info = linfo.get();
 
     LOG_INFO << "loading: " << model_id << ":" << version;
+    model_info->load_start_ns_ = now_ns;
```
[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:474]: load_start_ns_ exists without RITON_ENABLE_METRICS being enabled also. It just stores the time model loading was initiated in the triton.
Did not seem logical to add ifdef for one line.
The other metrics stuff is guarded by ifdef

File: src/model_repository_manager/model_lifecycle.cc [hunk 10]
```diff
@@ -799,10 +801,20 @@ ModelLifeCycle::OnLoadFinal(
     // Mark current versions ready and track info in foreground
     for (auto& loaded : load_tracker->load_set_) {
       std::lock_guard<std::mutex> curr_info_lk(loaded.second->mtx_);
-
       loaded.second->state_ = ModelReadyState::READY;
       loaded.second->state_reason_.clear();
-
+#ifdef TRITON_ENABLE_METRICS
+      auto reporter = loaded.second->model_->MetricReporter();
```
[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:807]: I tried to stay consistent with existing Reporting functions which are quite simple and devoid of any real logic just deal with `std::shared_ptr<MetricModelReporter> reporter,` and the value of the metric

File: src/metrics.cc (hunk 6 above)
Comment (indrajit96) [reply src/metrics.cc]: Resolved after ack from customer.

**Comment by kthui** [comment]
If the server is started with two models, the first model takes 0.1 second to load and the second model takes 10 seconds to load. I wonder what will the model_load_time metric report for each of the two models?

File: src/model_repository_manager/model_lifecycle.cc [hunk 11]
```diff
@@ -799,10 +801,11 @@ ModelLifeCycle::OnLoadFinal(
     // Mark current versions ready and track info in foreground
     for (auto& loaded : load_tracker->load_set_) {
       std::lock_guard<std::mutex> curr_info_lk(loaded.second->mtx_);
-
       loaded.second->state_ = ModelReadyState::READY;
       loaded.second->state_reason_.clear();
-
+#ifdef TRITON_ENABLE_METRICS
+      CalculateAndReportLoadTime(loaded.second);
+#endif  // TRITON_ENABLE_METRICS
```
[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:808]: This metric is per model so 0.2 and 10 seconds would be reported and stored for each of the model
https://github.com/triton-inference-server/core/pull/397/files#diff-d459d645388f914acbe22d1d7b9b689e2a1a349b2ab899df4c88de66c21235a0R474
We set the load_start_ns_ per model ONLY when it starts LOADING and calculated when model is READY
Below is example with multiple models loaded at startup

`# HELP nv_model_load_time Load Time per-model in seconds
# TYPE nv_model_load_time gauge
nv_model_load_time{model="simple_string",version="1"} 0.790762448
nv_model_load_time{model="simple_sequence",version="1"} 0.7909725990000001
nv_model_load_time{model="densenet_onnx",version="1"} 0.964002791
nv_model_load_time{model="simple_int8",version="1"} 0.775782746
nv_model_load_time{model="simple_identity",version="1"} 0.728239961
nv_model_load_time{model="simple",version="1"} 0.7150264079999999
nv_model_load_time{model="simple_dyna_sequence",version="1"} 0.730162601
nv_model_load_time{model="inception_graphdef",version="1"} 0.929637874`

Did I miss something?

[the remark this answers is not in the document]
Comment (kthui) [reply src/model_repository_manager/model_lifecycle.cc:808]: My bad, I mean across versions of the same model.

[the remark this answers is not in the document]
Comment (kthui) [reply src/model_repository_manager/model_lifecycle.cc:808]:
```
model_repo
|
`-- model_dir
    |
    `-- 1
    |   |
    |   `-- model.py   <-- takes 0.2 seconds to load
    |
    `-- 2
    |   |
    |   `-- model.py  <-- takes 10 seconds to load
    |
    `-- config.pbtxt
```
If both versions are loaded at the same time, is the metrics going to show
```
nv_model_load_time{model="model_name",version="1"} 0.2
nv_model_load_time{model="model_name",version="2"} 10
```

[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:808]: Yes, because we load versions sequentially
OnLoadComplete is where we set the metric
https://github.com/triton-inference-server/core/blob/571c5dcee325a78cd8f02e2ee8ebb17e1df2ac75/src/model_repository_manager/model_lifecycle.cc#L505
load_start_ns_ is set again for every version
https://github.com/triton-inference-server/core/blob/571c5dcee325a78cd8f02e2ee8ebb17e1df2ac75/src/model_repository_manager/model_lifecycle.cc#L467
We set load_start_ns_ in this loop ensuring this case.
Will post a local test result here

[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:808]: Below is the output of the config you explained

`nv_model_load_time{model="simple_string",version="1"} 0.282414657
nv_model_load_time{model="simple_sequence",version="1"} 0.28993499
nv_model_load_time{model="simple_identity",version="1"} 0.262991678
nv_model_load_time{model="densenet_onnx",version="1"} 0.352527317
nv_model_load_time{model="simple",version="1"} 0.216015252
nv_model_load_time{model="inception_graphdef",version="1"} 0.362119909
nv_model_load_time{model="simple_dyna_sequence",version="1"} 0.233802043
nv_model_load_time{model="simple_int8",version="1"} 0.265711744
nv_model_load_time{model="simple",version="2"} 0.216017807`

You can see different load times for `simple` model with different versions

[the remark this answers is not in the document]
Comment (kthui) [reply src/model_repository_manager/model_lifecycle.cc:808]: What happen if you add a 10 seconds delay to 
```
nv_model_load_time{model="simple",version="1"} 0.216015252
```
model version? Does it not affect
```
nv_model_load_time{model="simple",version="2"} 0.216017807
```
?

[the remark this answers is not in the document]
Comment (rmccorm4) [reply src/model_repository_manager/model_lifecycle.cc:813]: >  we load versions sequentially

FYI I believe the versions can be loaded in parallel: https://github.com/triton-inference-server/core/blob/c59250656e28636db9df079bf4693e382bda3c55/src/model_repository_manager/model_lifecycle.cc#L532-L534

[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:813]: Yes, I tested this particular scenario turns out due to async load subsequent versions of the model have in-correct timing even though in reality the time shoud be reflected only in version 1.

`nv_model_load_time{model="simple",version="1"} 10.216015252
nv_model_load_time{model="simple",version="2"} 10.216017807`

Syncing up with @kthui in sometime for this

File: src/model_repository_manager/model_lifecycle.h [hunk 12]
```diff
@@ -313,6 +314,12 @@ class ModelLifeCycle {
       ModelInfo* model_info, const bool is_update,
       const std::function<void(Status)>& OnComplete,
       std::shared_ptr<LoadTracker> load_tracker);
+  // Calculate time to load model
+  void CalculateAndReportLoadTime(ModelInfo* loaded_model_info);
```
Comment (yinggeh) [inline src/model_repository_manager/model_lifecycle.h:318]: #ifdef TRITON_ENABLE_METRICS

File: src/model_repository_manager/model_lifecycle.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (kthui) [reply src/model_repository_manager/model_lifecycle.cc:813]: Synced offline, if the `nv_model_load_time` metrics only care about the the duration for loading a new model (version), then we can simply time the duration for executing [this block](https://github.com/triton-inference-server/core/blob/main/src/backend_model.cc#L68-L232) of `TritonModel::Create()` code. We could have a separate metrics (i.e. `nv_model_update_time`) for tracking the update duration of a loaded model (version) by timing [this block](https://github.com/triton-inference-server/core/blob/main/src/backend_model.cc#L239-L270) of `TritonModel::UpdateInstanceGroup()` code.

**Comment by yinggeh** [comment]
Can you build with metrics disabled flags and make sure it works?
`-DTRITON_ENABLE_METRICS=OFF -DTRITON_ENABLE_METRICS_CPU=OFF -DTRITON_ENABLE_METRICS_GPU=OFF -DTRITON_ENABLE_STATS=OFF`

File: src/model_repository_manager/model_lifecycle.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc]: Fixed and tested as per our discussion.
` HELP nv_model_load_duration_secs Model load time in seconds
# TYPE nv_model_load_duration_secs gauge
nv_model_load_duration_secs{model="input_all_optional",version="2"} 1.529608532
nv_model_load_duration_secs{model="input_all_optional",version="1"} 11.699666445`

File: src/model_repository_manager/model_lifecycle.h (hunk 12 above)
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.h]: I don't see any metric reporting functions in .h being guarded by `#ifdef TRITON_ENABLE_METRICS` any reason to change the convention?
See `IncrementPendingRequestCount` and `DecrementPendingRequestCount`

Comment (yinggeh) [reply src/model_repository_manager/model_lifecycle.h]: Then that's fine.

File: src/model_repository_manager/model_lifecycle.cc [hunk 13]
```diff
@@ -847,4 +850,32 @@ ModelLifeCycle::OnLoadFinal(
   }
 }
 
+void
+ModelLifeCycle::CalculateAndReportLoadTime(ModelInfo* loaded_model_info)
+{
+  auto reporter = loaded_model_info->model_->MetricReporter();
```
Comment (yinggeh) [inline src/model_repository_manager/model_lifecycle.cc]: Why isn't this function guarded by `#ifdef TRITON_ENABLE_METRICS`?

File: src/model_repository_manager/model_lifecycle.cc [hunk 14]
```diff
@@ -847,4 +856,33 @@ ModelLifeCycle::OnLoadFinal(
   }
 }
 
+void
+ModelLifeCycle::CalculateAndReportLoadTime(
+    ModelInfo* loaded_model_info, std::unique_ptr<Model>* model)
+{
+  auto reporter = (*model)->MetricReporter();
```
[the remark this answers is not in the document]
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc:864]: Resolved had initially only guarded
ReportModelLoadTime as that reports the metrics.
Added guard for `CalculateAndReportLoadTime` too

**Review by yinggeh — DISMISSED** [review]
LGTM. Make sure you addressed all the comments before merge.

File: src/model_repository_manager/model_lifecycle.cc [hunk 15]
```diff
@@ -559,11 +560,21 @@ ModelLifeCycle::CreateModel(
...
     status = TritonModel::Create(
         server_, model_info->model_path_, options_.backend_cmdline_config_map,
         options_.host_policy_map, model_id, version, model_config,
         is_config_provided, &model);
     is.reset(model.release());
+    if (status.IsOk()) {
+#ifdef TRITON_ENABLE_METRICS
+      CalculateAndReportLoadTime(model_info, &is);
+#endif  // TRITON_ENABLE_METRICS
+    }
```
Comment (kthui) [inline src/model_repository_manager/model_lifecycle.cc:577]: If the `load_start_ns` is only for reporting load time, how about we simply pass the `load_start_ns` directly into `CalculateAndReportLoadTime()` function? i.e.
```c++
CalculateAndReportLoadTime(model_load_ns, &is);
```
without storing it in `model_info->load_start_ns_` first, because it will not be accessed again after reporting the load time.

**Review by kthui — DISMISSED** [review]
A minor comment on `model_info->load_start_ns_` lifecycle. LGTM overall!

File: src/model_repository_manager/model_lifecycle.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (kthui) [reply src/model_repository_manager/model_lifecycle.cc]: Nice job fixing the logic issue!

File: src/model_repository_manager/model_lifecycle.cc (hunk 15 above)
Comment (indrajit96) [reply src/model_repository_manager/model_lifecycle.cc]: Yes makes sense now as we pass model pointer for metrics reporter sperately.

**Review by kthui — APPROVED** [review]
